url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/grammar/SQLGrammarParser.java

```diff
@@ -0,0 +1,3083 @@
+		public TerminalNode LEFT_PARENTHESIS() { return getToken(SQLGrammarParser.LEFT_PARENTHESIS, 0); }
+		public TerminalNode RIGHT_PARENTHESIS() { return getToken(SQLGrammarParser.RIGHT_PARENTHESIS, 0); }
+		public TerminalNode HASH() { return getToken(SQLGrammarParser.HASH, 0); }
+		public CreateIndexContext(ParserRuleContext parent, int invokingState) {
+			super(parent, invokingState);
+		}
+		@Override public int getRuleIndex() { return RULE_createIndex; }
+		@Override
+		public void enterRule(ParseTreeListener listener) {
+			if ( listener instanceof SQLGrammarListener ) ((SQLGrammarListener)listener).enterCreateIndex(this);
+		}
+		@Override
+		public void exitRule(ParseTreeListener listener) {
+			if ( listener instanceof SQLGrammarListener ) ((SQLGrammarListener)listener).exitCreateIndex(this);
+		}
+	}
+
+	public final CreateIndexContext createIndex() throws RecognitionException {
+		CreateIndexContext _localctx = new CreateIndexContext(_ctx, getState());
+		enterRule(_localctx, 24, RULE_createIndex);
+
+		    boolean isUnique = false;
+		    String indexName = null;
+		    String tableName = null;
+		    List<String> attributeNames = new ArrayList<>();
+		    boolean isAscending = true;
+		    boolean isUsingHash = true;
+
+		int _la;
+		try {
+			enterOuterAlt(_localctx, 1);
+			{
+			setState(215); match(CREATE);
+			setState(218);
+			_la = _input.LA(1);
+			if (_la==UNIQUE) {
+				{
+				setState(216); match(UNIQUE);
+
+				        isUnique = true;
+				    
+				}
+			}
+
+			setState(220); match(INDEX);
+			setState(221); ((CreateIndexContext)_localctx).firstLevelId = firstLevelId();
+			setState(222); match(ON);
+
+			        indexName = ((CreateIndexContext)_localctx).firstLevelId.result;
+			    
+			setState(224); ((CreateIndexContext)_localctx).firstLevelId = firstLevelId();
+			setState(225); match(LEFT_PARENTHESIS);
+
+			        tableName = ((CreateIndexContext)_localctx).firstLevelId.result;
+			    
+			setState(227); ((CreateIndexContext)_localctx).firstLevelId = firstLevelId();
+
+			        attributeNames.add(((CreateIndexContext)_localctx).firstLevelId.result);
+			    
+			setState(232);
+			switch (_input.LA(1)) {
+			case ASC:
+				{
+				setState(229); match(ASC);
+				}
+				break;
+			case DESC:
+				{
+				{
+				setState(230); match(DESC);
+
+				        isAscending = false;
+				    
+				}
+				}
+				break;
+			case COMMA:
+			case RIGHT_PARENTHESIS:
+				break;
+			default:
+				throw new NoViableAltException(this);
+			}
+			setState(240);
+			_errHandler.sync(this);
+			_la = _input.LA(1);
+			while (_la==COMMA) {
+				{
+				{
+				setState(234); match(COMMA);
+				setState(235); ((CreateIndexContext)_localctx).firstLevelId = firstLevelId();
+
+				        attributeNames.add(((CreateIndexContext)_localctx).firstLevelId.result);
+				    
+				}
+				}
+				setState(242);
+				_errHandler.sync(this);
+				_la = _input.LA(1);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1112:1125, 1306:1319, 1623:1636.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.